Sources_v3/Workers/Background/MissingEventsReplayer.swift

```diff
@@ -0,0 +1,100 @@
+//
+
+import CoreData
+import Foundation
+
+/// The type is designed to obtain missing events that happened in watched channels while user
+/// was not connected to the web-socket.
+///
+/// The object listenes for `ConnectionStatusUpdated` events
+/// and remembers the `CurrentUserDTO.lastReceivedEventDate` when status becomes `connecting`.
+///
+/// When the status becomes `connected` the `/sync` endpoint is called
+/// with `lastReceivedEventDate` and `cids` of watched channels.
+///
+/// We remember `lastReceivedEventDate` when state becomes `connecting` to catch the last event date
+/// before the `HealthCheck` override the `lastReceivedEventDate` with the recent date.
+///
+class MissingEventsReplayer<ExtraData: ExtraDataTypes>: Worker {
+    // MARK: - Properties
+    
+    private var connectionObserver: EventObserver?
+    private var lastSyncedAt: Date?
+    
+    // MARK: - Init
+
+    override init(database: DatabaseContainer, webSocketClient: WebSocketClient, apiClient: APIClient) {
+        super.init(database: database, webSocketClient: webSocketClient, apiClient: apiClient)
+        startObserving()
+    }
+    
+    // MARK: - Private
+    
+    private func startObserving() {
+        connectionObserver = EventObserver(
+            notificationCenter: webSocketClient.eventNotificationCenter,
+            tranform: { $0 as? ConnectionStatusUpdated },
+            callback: { [unowned self] in
+                switch $0.connectionStatus {
+                case .connecting:
+                    self.obtainLastSyncDate()
+                case .connected:
+                    self.fetchAndReplayMissingEvents()
+                default:
+                    break
+                }
+            }
+        )
+    }
+    
+    private func obtainLastSyncDate() {
+        lastSyncedAt = database.backgroundReadOnlyContext.currentUser()?.lastReceivedEventDate
+    }
+    
+    private func fetchAndReplayMissingEvents() {
+        guard let lastSyncedAt = lastSyncedAt else { return }
+        
+        let watchedChannelIDs = channels.filter { $0.isWatched }.map(\.cid)
+        
+        let endpoint: Endpoint<MissingEventsPayload<ExtraData>> = .missingEvents(
+            since: lastSyncedAt,
+            cids: watchedChannelIDs
+        )
+        
+        apiClient.request(endpoint: endpoint) { [weak self] in
+            switch $0 {
+            case let .success(payload):
+                self?.webSocketClient.eventNotificationCenter.process(payload.eventPayloads)
+            case let .failure(error):
+                log.error("Internal error: Failed to fetch and reply missing events: \(error)")
+            }
+        }
+    }
+    
+    private var channels: [ChannelModel<ExtraData>] {
+        do {
+            let request = NSFetchRequest<ChannelDTO>(entityName: ChannelDTO.entityName)
+            let channelDTOs = try database.backgroundReadOnlyContext.fetch(request)
+            return channelDTOs.map(ChannelModel.create)
```

Comment:
Can you please add a comment that you really want to fetch all channels, or, alternatively, create a `ChannelDTO.allChannelsFetchRequest` helper? It looks like you forgot to specify what you want to fetch a little bit, so our future selfs will appreciate the comment that this was on purpose :)